pkg/controllers/jobset_controller.go

```diff
@@ -364,31 +369,23 @@ func (r *JobSetReconciler) restartPolicyRecreateAll(ctx context.Context, js *job
 	return nil
 }
 
-func (r *JobSetReconciler) deleteJobs(ctx context.Context, js *jobset.JobSet, jobsForDeletion []*batchv1.Job) error {
+func (r *JobSetReconciler) deleteJobs(ctx context.Context, jobsForDeletion []*batchv1.Job) error {
 	log := ctrl.LoggerFrom(ctx)
-
-	var wg sync.WaitGroup
+	lock := &sync.Mutex{}
 	var finalErr error
-
-	// Delete all jobs in parallel.
-	// TODO: limit number of goroutines used here.
-	for _, job := range jobsForDeletion {
-		job := job
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-
-			// Delete job. This deletion event will trigger another reconcilliation,
-			// where the jobs are recreated.
-			backgroundPolicy := metav1.DeletePropagationBackground
-			if err := r.Delete(ctx, job, &client.DeleteOptions{PropagationPolicy: &backgroundPolicy}); client.IgnoreNotFound(err) != nil {
-				finalErr = err
-				return
-			}
-			log.V(2).Info("successfully deleted job", "job", klog.KObj(job), "restart attempt", job.Labels[job.Labels[RestartsKey]])
-		}()
-	}
-	wg.Wait()
+	workqueue.ParallelizeUntil(ctx, parallelDeletions, len(jobsForDeletion), func(i int) {
+		targetJob := jobsForDeletion[i]
+		// Delete job. This deletion event will trigger another reconciliation,
+		// where the jobs are recreated.
+		backgroundPolicy := metav1.DeletePropagationBackground
+		if err := r.Delete(ctx, targetJob, &client.DeleteOptions{PropagationPolicy: &backgroundPolicy}); client.IgnoreNotFound(err) != nil {
+			lock.Lock()
```

Comment:
Why are you only acquiring the lock if an error occurs deleting the job (ignoring 'not found' errors)?

Reply:
Because this lock is only for `finalErr`.

Let me know if I'm missing anything.

Reply:
Ah I see now. Thanks for clarifying.